# Preambulo

**CONTROL DE LECTURA 3** · Duración: 20 minutos · Modalidad: individual · Entrega: notebook (.ipynb) o script (.py)

**Nombre:** ______________________________  **Fecha:** ______________________________

**Objetivo.** Contrastar dos gradientes candidatos con diferencias centrales e interpretar el efecto del paso y de la precisión numérica.

## Verificación numérica de gradientes mediante diferencias finitas

En este control se contrastan dos gradientes candidatos, `gA` y `gB`, contra una aproximación numérica de la derivada de `f` obtenida por **diferencias centrales**, `(f(w+h) - f(w-h)) / (2h)`.

La idea es determinar cuál candidato es compatible con la evidencia numérica en `w = 2` **sin derivar a mano ni entrenar**, y analizar cómo el tamaño del paso `h` y la precisión numérica (aritmética de punto flotante) afectan la calidad de la aproximación.

## Contexto

Se proporcionan una función y dos candidatos a su derivada:

- f(w) = w³ - 2w
- gA(w) = 3w² - 2
- gB(w) = 3w - 2

La tarea es determinar cuál candidato es compatible con la evidencia numérica en w = 2, sin derivar a mano ni entrenar.

## Tareas

El control consta de tres tareas: (1) **registrar una predicción**, respondiendo P0 **antes de ejecutar** y conservándola aunque después deba corregirse; (2) **completar una función**: solo la expresión pendiente de `diferencia_central(f, w, h)`, con el método central estudiado y sin usar la derivada analítica ni los candidatos; y (3) **ejecutar y observar** los casos con `w = 2.0` y `h = 1e-1`, `1e-5`, `1e-16` en float de Python, sin redondear cálculos, observando aproximaciones, candidatos, discrepancias absolutas, puntos perturbados y su igualdad real.

**Tarea 1 — Registrar una predicción (P0, antes de ejecutar)**

Antes de ejecutar cualquier cálculo se registra la siguiente predicción, que **se conservará** aunque después deba corregirse:

**(a) ¿Qué dos cantidades se comparan al verificar un gradiente?** Se comparan la **aproximación numérica de f'(w) por diferencias centrales**, es decir `(f(w+h) - f(w-h)) / (2h)`, y el **valor del gradiente candidato g(w)** en el mismo punto; la verificación consiste en comprobar que ambas coinciden dentro de una tolerancia acorde con el paso y la precisión de la máquina.

**(b) ¿Un h menor siempre mejora la aproximación?** **No necesariamente.** El error de truncamiento del método central disminuye al reducir h (escala como h²), pero con un h demasiado pequeño la resta `f(w+h) - f(w-h)` involucra números casi iguales y sufre **cancelación/absorción en punto flotante**: los puntos perturbados pueden volverse indistinguibles en la precisión disponible y el error de redondeo llega a dominar o incluso anular la diferencia. Existe, por tanto, un **h óptimo intermedio**: reducir h mejora la aproximación solo hasta cierto punto.

**Tarea 2 y Tarea 3 — Completar la función y ejecutar la evidencia**

Se completa únicamente la expresión pendiente de `diferencia_central(f, w, h)` con el **método central** estudiado; la función no usa la derivada analítica ni los candidatos. Se ejecutan los tres casos con `w = 2.0` y `h = 1e-1`, `1e-5`, `1e-16`, todo en float de Python y **sin redondear los cálculos** (solo se formatea la impresión). Por cada caso se muestran: la aproximación central, `gA(2.0)`, `gB(2.0)`, las discrepancias absolutas `|aprox - gA(2.0)|` y `|aprox - gB(2.0)|`, los puntos perturbados `w+h` y `w-h`, y su igualdad real en float (`w+h == w-h`). Los resultados se guardan en `resultados_control_3.csv` (una fila por `h`).

In [1]:
# -*- coding: utf-8 -*-
"""
SUBTASK s2 - Verificacion numerica de gradientes mediante diferencias centrales.

f(w)  = w**3 - 2*w
gA(w) = 3*w**2 - 2   (candidato A)
gB(w) = 3*w - 2      (candidato B)

Con w = 2.0 y h en [1e-1, 1e-5, 1e-16], todo en float de Python y sin redondear
calculos (solo se formatea la impresion), se calcula por caso:
aproximacion central, gA(2.0), gB(2.0), discrepancias absolutas |aprox-gA| y
|aprox-gB|, puntos perturbados w+h y w-h, y su igualdad real en float.

Salidas:
  - tabla en stdout con las tres filas
  - resultados_control_3.csv (una fila por h)
  - resultados.json (todos los numeros calculados)
"""

import csv
import json


def f(w):
    """Funcion objetivo: f(w) = w**3 - 2*w."""
    return w**3 - 2*w


def gA(w):
    """Candidato A al gradiente: gA(w) = 3*w**2 - 2."""
    return 3*w**2 - 2


def gB(w):
    """Candidato B al gradiente: gB(w) = 3*w - 2."""
    return 3*w - 2


def diferencia_central(f, w, h):
    """Aproximacion de f'(w) por diferencias centrales:
       (f(w+h) - f(w-h)) / (2*h).
       No usa la derivada analitica ni los candidatos gA/gB."""
    return (f(w + h) - f(w - h)) / (2 * h)


def fmt(v):
    """Formato de impresion/escritura: repr de float (sin redondeo, round-trip)."""
    if isinstance(v, bool):
        return str(v)
    return repr(float(v))


def main():
    w = 2.0
    pasos = [1e-1, 1e-5, 1e-16]

    filas = []
    for h in pasos:
        aprox = diferencia_central(f, w, h)      # float de Python, sin redondear
        valor_gA = gA(w)                          # gA(2.0) = 10.0
        valor_gB = gB(w)                          # gB(2.0) = 4.0
        disc_gA = abs(aprox - valor_gA)
        disc_gB = abs(aprox - valor_gB)
        w_mas_h = w + h
        w_menos_h = w - h
        iguales = (w_mas_h == w_menos_h)          # igualdad real en float
        filas.append({
            "h": h,
            "aproximacion": aprox,
            "gA": valor_gA,
            "gB": valor_gB,
            "disc_abs_gA": disc_gA,
            "disc_abs_gB": disc_gB,
            "w_mas_h": w_mas_h,
            "w_menos_h": w_menos_h,
            "perturbados_iguales": iguales,
        })

    # ---------- tabla en stdout ----------
    columnas = ["h", "aproximacion", "gA", "gB", "disc_abs_gA",
                "disc_abs_gB", "w_mas_h", "w_menos_h", "perturbados_iguales"]
    tabla = [[fmt(fila[c]) for c in columnas] for fila in filas]
    anchos = [max([len(c)] + [len(t[i]) for t in tabla])
              for i, c in enumerate(columnas)]

    def linea(celdas):
        return " | ".join(c.ljust(a) for c, a in zip(celdas, anchos))

    print("Verificacion numerica de gradientes por diferencias centrales")
    print("f(w) = w**3 - 2*w | gA(w) = 3*w**2 - 2 | gB(w) = 3*w - 2")
    print("w = 2.0, float de Python, calculos sin redondear")
    print()
    print(linea(columnas))
    print("-+-".join("-" * a for a in anchos))
    for t in tabla:
        print(linea(t))
    print()

    # ---------- CSV: resultados_control_3.csv ----------
    with open("resultados_control_3.csv", "w", newline="") as fh:
        writer = csv.writer(fh)
        writer.writerow(columnas)
        for fila in filas:
            writer.writerow([fmt(fila[c]) for c in columnas])

    # ---------- JSON: resultados.json ----------
    resultados = {
        "descripcion": "Verificacion numerica de gradientes por diferencias centrales",
        "f": "w**3 - 2*w",
        "gA": "3*w**2 - 2",
        "gB": "3*w - 2",
        "w": w,
        "casos": [
            {
                "h": fila["h"],
                "aproximacion": fila["aproximacion"],
                "gA": fila["gA"],
                "gB": fila["gB"],
                "disc_abs_gA": fila["disc_abs_gA"],
                "disc_abs_gB": fila["disc_abs_gB"],
                "w_mas_h": fila["w_mas_h"],
                "w_menos_h": fila["w_menos_h"],
                "perturbados_iguales": fila["perturbados_iguales"],
            }
            for fila in filas
        ],
    }
    with open("resultados.json", "w") as fh:
        json.dump(resultados, fh, indent=2)

    print("Archivos escritos: resultados_control_3.csv, resultados.json")


if __name__ == "__main__":
    main()


Verificacion numerica de gradientes por diferencias centrales
f(w) = w**3 - 2*w | gA(w) = 3*w**2 - 2 | gB(w) = 3*w - 2
w = 2.0, float de Python, calculos sin redondear

h     | aproximacion       | gA   | gB  | disc_abs_gA            | disc_abs_gB       | w_mas_h | w_menos_h | perturbados_iguales
------+--------------------+------+-----+------------------------+-------------------+---------+-----------+--------------------
0.1   | 10.010000000000007 | 10.0 | 4.0 | 0.010000000000006892   | 6.010000000000007 | 2.1     | 1.9       | False              
1e-05 | 10.000000000198739 | 10.0 | 4.0 | 1.9873880319210002e-10 | 6.000000000198739 | 2.00001 | 1.99999   | False              
1e-16 | 0.0                | 10.0 | 4.0 | 10.0                   | 4.0               | 2.0     | 2.0       | True               

Archivos escritos: resultados_control_3.csv, resultados.json


## Preguntas de interpretación

A continuación se responden P0 a P3: P0 conserva la predicción registrada en la Tarea 1 antes de ejecutar, y P1 a P3 interpretan la evidencia numérica de la tabla de salida anterior y de `resultados_control_3.csv`.

**P0 — Predicción registrada antes de ejecutar (se conserva).** La predicción registrada en la Tarea 1 se mantiene sin cambios: (a) al verificar un gradiente se comparan la **aproximación de f'(w) por diferencias centrales**, `(f(w+h) - f(w-h)) / (2h)`, y el **valor del candidato g(w)** en el mismo punto; (b) **un h menor no siempre mejora la aproximación**: el error de truncamiento disminuye con h, pero con un h demasiado pequeño la resta `f(w+h) - f(w-h)` sufre cancelación/absorción en aritmética de punto flotante y el error puede crecer; existe un h óptimo intermedio.

**P1 — ¿Qué candidato respalda una fila fiable?** La fila fiable es la del paso intermedio, h = 1e-5. Como muestra la salida anterior, en esa fila la aproximación central es ≈ 10.0000000002, con discrepancia absoluta ≈ 2×10⁻¹⁰ frente a gA y ≈ 6.0 frente a gB: la aproximación coincide con gA(2.0) hasta el orden de 10⁻¹⁰ y dista ≈ 6 de gB(2.0). (La fila de h = 1e-1 ya apuntaba a gA, con discrepancia del orden de 10⁻² por truncamiento, pero es menos precisa.) Esta evidencia, observada en la fila numéricamente confiable, **respalda al candidato gA** como derivada de f en w = 2 y descarta a gB.

**P2 — Comparación entre el menor h y el h intermedio.** Con el h intermedio (1e-5) la aproximación es ≈ 10.0000000002 y los puntos perturbados sí difieren: w+h = 2.00001 y w-h = 1.99999, con `perturbados_iguales = False`. Con el menor (1e-16) la aproximación colapsa a 0.0 y los puntos dejan de distinguirse: w+h = w-h = 2.0, con `perturbados_iguales = True`. **Causa:** en doble precisión, sumar 1e-16 a 2.0 se absorbe (está por debajo del épsilon de máquina relativo a 2.0, ≈ 2.2×10⁻¹⁶), de modo que f(w+h) y f(w-h) se evalúan en el mismo punto y la diferencia central da exactamente cero: el error de redondeo domina por completo sobre el de truncamiento. **Contraste con P0:** el experimento confirma la predicción — reducir h baja el error de truncamiento solo hasta cierto punto; después, el error de redondeo por cancelación/absorción crece y arruina la estimación, de modo que el mejor h es intermedio.

**P3 — ¿Coincidir en w = 2 demuestra corrección para cualquier entrada?** **No.** La evidencia en un solo punto verifica el **valor** del gradiente en w = 2, no la **forma funcional** del candidato: dos funciones distintas pueden coincidir en puntos aislados, de modo que la coincidencia en w = 2 no se extiende a cualquier entrada. **Otra comprobación (sin programarla):** repetir el mismo contraste con diferencias centrales en otros puntos, por ejemplo w = 0, w = -1 y w = 0.5, y verificar que el candidato elegido sigue coincidiendo con la aproximación en todos ellos; alternativamente, contrastar contra la derivada simbólica en una malla de puntos, o repetir el experimento con otra f y sus candidatos. **Cierre sobre P0:** la predicción registrada antes de ejecutar **se confirmó**: se comparaban la aproximación central con el valor de cada candidato, y el caso del paso más pequeño mostró que un h menor no mejora siempre la aproximación.

## Entrega

**Entregables de este control:**

1. **Notebook ejecutado** con nombre (consignado en el preambulo), la función `diferencia_central` completada por método central —sin usar la derivada analítica ni los candidatos— y las respuestas **P0 a P3**.
2. **Evidencia numérica** de los tres casos con `w = 2.0` y `h = 1e-1`, `1e-5`, `1e-16`, en float de Python y sin redondear cálculos; los tres casos quedan conservados en la tabla de salida.
3. **Archivo `resultados_control_3.csv`** con las columnas `h, aproximacion, gA, gB, disc_abs_gA, disc_abs_gB, w_mas_h, w_menos_h, perturbados_iguales` (una fila por `h`), generado por la celda de código anterior, que además escribe `resultados.json` como respaldo.

**Estado de completitud:** el control fue realizado en su totalidad — predicción P0 registrada antes de ejecutar y conservada, función `diferencia_central` completada, evidencia de los tres casos obtenida y respuestas P0 a P3 redactadas. **No queda nada pendiente.**